processing/src/main/java/io/druid/segment/data/CompressedVSizeIntsIndexedSupplier.java

```diff
@@ -0,0 +1,423 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.data;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closeables;
+import com.google.common.primitives.Ints;
+import com.google.common.primitives.Shorts;
+import com.metamx.common.IAE;
+import com.metamx.common.guava.CloseQuietly;
+import io.druid.collections.ResourceHolder;
+import io.druid.collections.StupidResourceHolder;
+import io.druid.segment.CompressedPools;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.IntBuffer;
+import java.nio.ShortBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.util.Iterator;
+import java.util.List;
+
+public class CompressedVSizeIntsIndexedSupplier implements WritableSupplier<IndexedInts>
+{
+  public static final byte version = 0x2;
+
+  private final int totalSize;
+  private final int sizePer;
+  private final int numBytes;
+  private final int bigEndianShift;
+  private final int littleEndianMask;
+  private final GenericIndexed<ResourceHolder<ByteBuffer>> baseBuffers;
+  private final CompressedObjectStrategy.CompressionStrategy compression;
+
+  CompressedVSizeIntsIndexedSupplier(
+      int totalSize,
+      int sizePer,
+      int numBytes,
+      GenericIndexed<ResourceHolder<ByteBuffer>> baseBuffers,
+      CompressedObjectStrategy.CompressionStrategy compression
+  )
+  {
+    this.totalSize = totalSize;
+    this.sizePer = sizePer;
+    this.baseBuffers = baseBuffers;
+    this.compression = compression;
+    this.numBytes = numBytes;
+    this.bigEndianShift = Integer.SIZE - (numBytes << 3); // numBytes * 8
+    this.littleEndianMask = (int)((1L << (numBytes << 3)) - 1); // set numBytes * 8 lower bits to 1
+  }
+
+  public static int maxIntsInBufferForBytes(int numBytes)
+  {
+    return (CompressedPools.BUFFER_SIZE - bufferPadding(numBytes)) / numBytes;
+  }
+
+  private static int bufferPadding(int numBytes)
+  {
+    // when numBytes == 3 we need to pad the buffer to allow reading an extra byte
+    // beyond the end of the last value, since we use buffer.getInt() to read values.
+    // for numBytes < 3 we remove the need for padding by reading short or bytes directly.
+    if (numBytes == 3) {
+      return Ints.BYTES - numBytes;
+    } else {
+      return 0;
+    }
+  }
+
+  public static int maxIntsInBufferForValue(int maxValue)
+  {
+    return maxIntsInBufferForBytes(VSizeIndexedInts.getNumBytesForMax(maxValue));
+  }
+
+  public int size()
+  {
+    return totalSize;
+  }
+
+  @Override
+  public IndexedInts get()
+  {
+    final boolean powerOf2 = sizePer == (1 << Integer.numberOfTrailingZeros(sizePer));
+
+    if(powerOf2) {
+      if(numBytes == Ints.BYTES) {
+        return new CompressedFullSizeIndexedInts() {
```

Comment:
Are you talking about the override for `get2` of the optimization for different numBytes?

For decompressed buffers with a power of two value count, we have been able to get a few percent extra speed by doing bit operations instead of divisions and remainder calculations to determine the index into the buffer for existing CompressedXXXIndexedSuppliers. We used to have non-power-of-2-sized buffers in previous versions, so I left it as an override for consistency. Arguably this is not necessary for numBytes = 3, since the buffer will never be able to hold a power of two count. As far as the call is concerned, this is called enough times that it gets inlined.

Regarding the optimization for different byte sizes, it removes the need for padding for numBytes other than 3. It also uses ShortBuffer and ByteBuffer directly, which hopefully makes reads more predictable by not having them overlap, and may benefit from JVM intrinsics if there are any.